### ISSUE TITLE
[HOLD for payment 2024-10-10] [$1000] Enable the noUncheckedIndexedAccess TS compiler option

### ISSUE THREAD

**Issue by roryabraham** [Expensify/App#43055]
Slack proposal: https://expensify.slack.com/archives/C01GTK53T8Q/p1717468504413819

### Problem

We currently have [crash occurring on production](https://github.com/Expensify/App/issues/42885). The problematic code can be summarized with a minimal example:

```ts
type MyType = {
    something: string;
};

const myArr: MyType[] = [];

// myItem is inferred to have type MyType, when really it is undefined 
const myItem = myArr[42];

// No compiler error, but the app will crash
console.log(myItem.something);
```

As you can see, we unsafely indexed an array, and then assumed the result was defined. Then trying to access a property of `undefined`, we experience a crash.

### Solution

Enable the [noUncheckedArrayAccess](https://www.typescriptlang.org/tsconfig/#noUncheckedIndexedAccess) TypeScript config. With that config enabled, the type of `myItem` is correctly inferred to `MyType | undefined`, and we get a compiler error when trying to access `myItem.something`, preventing the crash at compile time.

<details><summary>Issue Owner</summary>Current Issue Owner: @adelekennedy</details>

**Comment by roryabraham** [comment]
It's worth noting that when enabling this rule we get more than 1000 errors across more than 200 files, so in order for me to accept a proposal it would need to include a rollout plan so we can incrementally adopt this new rule across a subset of files at a time.

**Comment by eh2077** [comment]
@ShridharGoel Thanks for your proposal. 

How will the new `tsconfig.incremental.json` cooperate with `npm run typecheck` which uses `tsconfig.json` from the root directory? 
Is it possible to achieve the goal by editing the `tsconfig.json` from project root directory?

**Comment by roryabraham** [comment]
> We can use something like tsc -p tsconfig.uncheckedindex.json as the command (using uncheckedindex in the name now).

That sounds reasonable. So the rollout plan would look something like this:

1. Create a new `tsconfig.uncheckedindex.json` that only applies the rule to some reasonable subset of files.
1. Update `npm run typecheck` to do this:

  ```bash
  tsc && tsc -p tsconfig.uncheckedindex.json
  ```
1. Incrementally update `tsconfig.uncheckedindex.json`, module by module, until in includes all files
1. Add `noUncheckedArrayAccess` to the base `tsconfig.json`, remove `tsconfig.uncheckedindex.json` and restore `npm run typecheck` to the way it is now.

sounds good, the only other thing I'd like to figure out is how we want to break it up into reasonably-sized PRs. Some of these might be a bit trickier than your standard TS migration, because it actually involves changing runtime code.

Going to assign this one a bounty of $1000 because it's going to be a few PRs.

**Comment by roryabraham** [comment]
@ShridharGoel I look forward to reviewing the first PR.

**Comment by roryabraham** [comment]
Just as a heads up, @blazejkustra and @fabioh8010 ended up objecting to the addition of this rule. @fabioh8010 has an alternate suggestion that he'll propose - using `.at()` instead of `[]` to index arrays (since it has the same effect and gives us checked indexed access of arrays in particular.

we may want to pause here for further consensus

**Comment by fabioh8010** [comment]
Posted the proposal [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1718102555930349?thread_ts=1717468504.413819&cid=C01GTK53T8Q).

**Comment by fabioh8010** [comment]
@roryabraham Did you have a chance to look at the [.at() proposal](https://expensify.slack.com/archives/C01GTK53T8Q/p1718102555930349?thread_ts=1717468504.413819&cid=C01GTK53T8Q)?

**Comment by roryabraham** [comment]
Sorry for the delay here, let's switch to the `.at() proposal` 👍🏼

**Comment by fabioh8010** [comment]
Nice @ShridharGoel ! So the first step would build a rule ESLint rule to enforce the use of the Array `.at()` method instead of direct access with `array[]`, more details [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1718102555930349?thread_ts=1717468504.413819&cid=C01GTK53T8Q).

It's important that the ESLint rule should be able to not only identify `array[]` usage and suggest `.at()` instead, but also be able to auto-fix those situations. The rule should be created on https://github.com/Expensify/eslint-config-expensify.

Please let us know if you have any further questions, thanks!

**Comment by fabioh8010** [comment]
@ShridharGoel I tried to review and test today but I'm having some local issues with the tests, I will let you know once I'm able to fully review the PR.

**Comment by fabioh8010** [comment]
The [ESLint PR](https://github.com/Expensify/eslint-config-expensify/pull/104) was merged! @ShridharGoel Do you have an ETA for the E/App PR?

**Comment by fabioh8010** [comment]
@ShridharGoel bump

**Comment by blazejkustra** [comment]
@ShridharGoel Friendly bump 🙇‍♂️

**Comment by eh2077** [comment]
@ShridharGoel Kindly comment an ETA for the E/App PR, thank you 🙇‍♂️

**Comment by roryabraham** [comment]
Merged https://github.com/Expensify/eslint-config-expensify/pull/114. Excited to complete https://github.com/Expensify/App/pull/47439

**Comment by roryabraham** [comment]
At this point, the last PR has been dragging on too long without enough proactive or responsive communication. As such, if it is not merged within the next one or two days, we are going to reassign this issue

**Comment by zfurtak** [comment]
Hello, I'm Zuza from swm agency and I'd like to work on this issue.

**Comment by zfurtak** [comment]
As it was decided in the [slack discussion](https://swmansion.slack.com/archives/C0593T50UHG/p1726771034164829?thread_ts=1726247566.017509&cid=C0593T50UHG) I will take over this issue

**Comment by roryabraham** [comment]
There are problems being reported with this rule where it's applying to other array methods such as `.slice` which should not be affected

**Comment by zfurtak** [comment]
So far, the problems I heard about (related to functions `.map(...)`, `.filter(..)` etc) appeared only in the IDE, and simply `rm -rf node_modules && npm i` fixed them. Isn't it a similar situation?

**Comment by roryabraham** [comment]
Yes, was just coming back to correct myself. Sorry about that https://expensify.slack.com/archives/C049HHMV9SM/p1728040189319779?thread_ts=1728038864.830499&cid=C049HHMV9SM

**Comment by adelekennedy** [comment]
@ShridharGoel & @eh2077 sent you offers in [Upwork](https://www.upwork.com/ab/applicants/1845656360323991155/offers)

[Job Posting](https://www.upwork.com/jobs/~021845656360323991155)

**Comment by eh2077** [comment]
@adelekennedy Thanks for checking payment of this task. I'm not sure if we're eligible for the full payment of this job because the task was reassigned in the middle of the last PR, see here https://github.com/Expensify/App/issues/43055#issuecomment-2353505683

I think it'll be better to wait for @roryabraham 's input before issuing the payment.

**Comment by adelekennedy** [comment]
Oh thank you @eh2077 I totally missed that (this issue has gotten pretty dense) Rory is out on parental leave so I'm making a call based on all the comments above - I think an even split here may be the most fair between you and @ShridharGoel?

**Comment by eh2077** [comment]
> Oh thank you @eh2077 I totally missed that (this issue has gotten pretty dense) Rory is out on parental leave so I'm making a call based on all the comments above - I think an even split here may be the most fair between you and @ShridharGoel?

@adelekennedy Personally, the payment of $500 is fair. Thanks!

**Comment by adelekennedy** [comment]
Yes!

**Comment by adelekennedy** [comment]
@eh2077 yes! I just withdrew the offer on Upwork and updated the payment summary above

**Comment by eh2077** [comment]
Thanks! Requested in NewDot.

**Comment by adelekennedy** [comment]
closing this for now - @ShridharGoel I will pay you once you accept in NewDot
